File: Dynamic Programming/213. House Robber II.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        if len(nums)== 0:
            return 0
        max1 = robHelper(nums[0:len(nums)-1])
        max2 = robHelper(nums[1:len(nums)+1])
        return max(max1,max2)
        
def robHelper(nums):
    print(nums)
    if len(nums) == 0:
        return 0
    if len(nums) ==1:
        return nums[0]
    if len(nums) == 2:
        return max(nums)
    maxRob = [0 for x in range(len(nums))]
    maxRob[0] = nums[0]
    maxRob[1] = nums[1]
    
    for robAmount in range(2,len(maxRob)):


        maxRob[robAmount] = max(maxRob[0:robAmount-1]) + nums[robAmount]

    return max(maxRob)

File: Dynamic Programming/213. House Robber II.cpp
class Solution {
public:
    int rob1(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, 0);
        if (n == 1) return nums[0];
        if (n == 2) {
            return *max_element(nums.begin(), nums.end());
        }
        dp[0] = nums[0];
        dp[1] = max(nums[0], nums[1]);

        for (int i = 2; i < n ; i++) {
            dp[i] = max(dp[i - 2] + nums[i], dp[i - 1]);
        }

        return *max_element(dp.begin(), dp.end());
    }


    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> nums1(n - 1 , 0);
        vector<int> nums2(n - 1, 0);
        if (n == 1) return nums[0];
        if (n == 2) {
            return max(nums[0], nums[1]);
        }
        for (int i = 0; i < n - 1; i++) {
            nums1[i] = nums[i];
        }
        for (int i = 1; i < n; i++) {
            nums2[i - 1] = nums[i];
        }
        int res1 = rob1(nums1);
        int res2 = rob1(nums2);
        return max(res1, res2);
    }
};